transformer_engine/pytorch/attention.py

```diff
@@ -481,91 +491,166 @@ def forward(ctx, q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k
+                                        causal=False, return_softmax=False,
+                                    )
+                    else:
+                        if use_fused_attention:
+                            out_per_step[i], [softmax_lse_per_step[i], rng_states[i]] = fused_attn_fwd(
+                                is_training, max_seqlen_q, max_seqlen_k, cu_seqlens_q, cu_seqlens_k,
+                                q, kv_inputs[i%2][0], kv_inputs[i%2][1], TE_DType[q.dtype],
+                                tex.NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen,
+                                attn_scale=softmax_scale, dropout=dropout_p,
+                                qkv_layout="bshd_bshd_bshd", attn_mask_type="no_mask",
+                            )
                         else:
-                            # [b, sq//2, np, hn] -> [b*sq//2, np, hn]
-                            q_inputs[i%2] = q[:, 1, ...].contiguous().view(-1, *q.shape[-2:])
-                            # [2, b, 2, sk//2, np, hn] -> [2, b*sk, np, hn]
+                            # [b, sq, np, hn] -> [b*sq, np, hn]
+                            q_inputs[i%2] = q.view(-1, *q.shape[-2:])
+                            # [2, b, sk, np, hn] -> [2, b*sk, np, hn]
                             kv_inputs[i%2] = kv_inputs[i%2].view(2, -1, *k.shape[-2:])
                             if _flash_attn_2_available:
                                 _, _, _, _, out_per_step[i], \
                                 softmax_lse_per_step[i], _, rng_states[i] = _flash_attn_forward(
                                     q_inputs[i%2], kv_inputs[i%2][0], kv_inputs[i%2][1],
-                                    cu_seqlens_q//2, cu_seqlens_k, max_seqlen_q//2, max_seqlen_k,
+                                    cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
                                     dropout_p, softmax_scale, causal=False, return_softmax=False,
+                                    **fa_optional_forward_kwargs
                                 )
                             else:
                                 out_per_step[i] = torch.empty_like(q_inputs[i%2])
                                 _, softmax_lse_per_step[i], rng_states[i], _ = _flash_attn_forward( # pylint: disable=unbalanced-tuple-unpacking
                                     q_inputs[i%2], kv_inputs[i%2][0], kv_inputs[i%2][1],
-                                    out_per_step[i], cu_seqlens_q//2, cu_seqlens_k,
-                                    max_seqlen_q//2, max_seqlen_k, dropout_p, softmax_scale,
+                                    out_per_step[i], cu_seqlens_q, cu_seqlens_k,
+                                    max_seqlen_q, max_seqlen_k, dropout_p, softmax_scale,
                                     causal=False, return_softmax=False,
                                 )
-                    else:
-                        assert False, "Not implemented yet!"
```

Comment:
Should we keep the `else` branch here? Do we support any other masks than `causal`?